File: monolith-spigot/src/main/java/gg/warcraft/monolith/spigot/event/SpigotItemEventMapper.java
package gg.warcraft.monolith.spigot.event;

import com.google.inject.Inject;
import gg.warcraft.monolith.api.core.EventService;
import gg.warcraft.monolith.api.entity.EntityType;
import gg.warcraft.monolith.api.item.event.ItemPickupEvent;
import gg.warcraft.monolith.api.item.event.ItemPrePickupEvent;
import gg.warcraft.monolith.api.world.item.Item;
import gg.warcraft.monolith.app.item.event.SimpleItemPickupEvent;
import gg.warcraft.monolith.app.item.event.SimpleItemPrePickupEvent;
import gg.warcraft.monolith.spigot.world.item.SpigotItemMapper;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityPickupItemEvent;

import java.util.UUID;

public class SpigotItemEventMapper implements Listener {
    private EventService eventService;
    private final SpigotItemMapper itemMapper;

    @Inject
    public SpigotItemEventMapper(EventService eventService,
                                 SpigotItemMapper itemMapper) {
        this.eventService = eventService;
        this.itemMapper = itemMapper;
    }

    @EventHandler(priority = EventPriority.HIGH)
    public void onEntityPickupItemEvent(EntityPickupItemEvent event) {
        UUID itemId = event.getItem().getUniqueId();
        Item item = itemMapper.map(event.getItem().getItemStack()).get(); // TODO use Option

        UUID entityId = event.getEntity().getUniqueId();
        EntityType entityType = EntityType.valueOf(event.getEntity().getType().name());

        ItemPrePickupEvent itemPrePickupEvent = new SimpleItemPrePickupEvent(itemId, item, entityId, entityType, event.isCancelled());
        eventService.publish(itemPrePickupEvent);

        event.setCancelled(!itemPrePickupEvent.isAllowed());
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    public void onEntityPickupItemEventMonitor(EntityPickupItemEvent event) {
        UUID itemId = event.getItem().getUniqueId();
        Item item = itemMapper.map(event.getItem().getItemStack()).get(); // TODO use Option

        UUID entityId = event.getEntity().getUniqueId();
        EntityType entityType = EntityType.valueOf(event.getEntity().getType().name());

        ItemPickupEvent itemPickupEvent = new SimpleItemPickupEvent(itemId, item, entityId, entityType);
        eventService.publish(itemPickupEvent);
    }
}

File: monolith-spigot/src/main/scala/gg/warcraft/monolith/spigot/world/item/SpigotItemEventMapper.scala
package gg.warcraft.monolith.spigot.world.item

import java.util.UUID

import gg.warcraft.monolith.api.core.EventService
import gg.warcraft.monolith.api.entity.EntityType
import gg.warcraft.monolith.api.world.item.{Item, ItemPickupEvent, ItemPrePickupEvent}
import org.bukkit.event.{EventHandler, EventPriority, Listener}

class SpigotItemEventMapper(
    implicit private val eventService: EventService,
    implicit private val itemMapper: SpigotItemMapper
) extends Listener {
  private def parse(
      item: SpigotItem,
      entity: SpigotEntity
  ): (UUID, Item, UUID, EntityType) = (
    item.getUniqueId,
    itemMapper.map(item.getItemStack).get,
    entity.getUniqueId,
    EntityType.valueOf(entity.getType.name)
  )

  @EventHandler(priority = EventPriority.HIGH)
  def prePickup(event: SpigotItemPickupEvent): Unit = {
    val (itemId, item, entityId, entityType) = parse(event.getItem, event.getEntity)
    val cancelled = event.isCancelled
    val prePickupEvent = ItemPrePickupEvent(
      item,
      itemId,
      entityId,
      entityType,
      cancelled
    )

    val reducedEvent = eventService.publish(prePickupEvent)
    event.setCancelled(!reducedEvent.allowed)
  }

  @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
  def onPickup(event: SpigotItemPickupEvent): Unit = {
    val (itemId, item, entityId, entityType) = parse(event.getItem, event.getEntity)
    val pickupEvent = ItemPickupEvent(item, itemId, entityId, entityType)

    eventService.publish(pickupEvent)
  }
}
